cmsplugin/cmsplugin.go

```diff
@@ -0,0 +1,447 @@
+package cmsplugin
+
+import (
+	"encoding/json"
+)
+
+type ErrorStatusT int
+
+// Plugin settings, kinda doesn;t go here but for now it is fine
+const (
+	Version               = "1"
+	ID                    = "cms"
+	CmdVoteDetails        = "votedccdetails"
+	CmdStartVote          = "startdccvote"
+	CmdCastVote           = "castdccvote"
+	CmdDCCVotes           = "dccvotes"
+	CmdInventory          = "cmsinventory"
+	CmdTokenInventory     = "dccinventory"
+	CmdVoteSummary        = "votedccsummary"
+	CmdLoadDCCVoteResults = "loaddccvoteresults"
+	MDStreamVoteBits      = 16 // Vote bits and mask
+	MDStreamVoteSnapshot  = 17 // Vote tickets and start/end parameters
+
+	VoteDurationMin = 2016 // Minimum vote duration (in blocks)
+	VoteDurationMax = 4032 // Maximum vote duration (in blocks)
+
+	// Error status codes
+	ErrorStatusInvalid          ErrorStatusT = 0
+	ErrorStatusInternalError    ErrorStatusT = 1
+	ErrorStatusDCCNotFound      ErrorStatusT = 2
+	ErrorStatusInvalidVoteBit   ErrorStatusT = 3
+	ErrorStatusVoteHasEnded     ErrorStatusT = 4
+	ErrorStatusDuplicateVote    ErrorStatusT = 5
+	ErrorStatusIneligibleUserID ErrorStatusT = 6
+)
+
+var (
+	// ErrorStatus converts error status codes to human readable text.
+	ErrorStatus = map[ErrorStatusT]string{
+		ErrorStatusInvalid:          "invalid error status",
+		ErrorStatusInternalError:    "internal error",
+		ErrorStatusDCCNotFound:      "dcc not found",
+		ErrorStatusInvalidVoteBit:   "invalid vote bit",
+		ErrorStatusVoteHasEnded:     "vote has ended",
+		ErrorStatusDuplicateVote:    "duplicate vote",
+		ErrorStatusIneligibleUserID: "inegligible user id",
+	}
+)
+
+// VoteOption describes a single vote option.
+type VoteOption struct {
+	Id          string `json:"id"`          // Single unique word identifying vote (e.g. yes)
+	Description string `json:"description"` // Longer description of the vote.
+	Bits        uint64 `json:"bits"`        // Bits used for this option
+}
+
+// Vote represents the vote options for vote that is identified by its token.
+type Vote struct {
+	Token            string       `json:"token"`            // Token that identifies vote
+	Mask             uint64       `json:"mask"`             // Valid votebits
+	Duration         uint32       `json:"duration"`         // Duration in blocks
+	QuorumPercentage uint32       `json:"quorumpercentage"` // Percent of eligible votes required for quorum
+	PassPercentage   uint32       `json:"passpercentage"`   // Percent of total votes required to pass
+	Options          []VoteOption `json:"options"`          // Vote option
+}
+
+// EncodeVote encodes Vote into a JSON byte slice.
+func EncodeVote(v Vote) ([]byte, error) {
+	return json.Marshal(v)
+}
+
+// DecodeVote decodes a JSON byte slice into a Vote.
+func DecodeVote(payload []byte) (*Vote, error) {
+	var v Vote
+
+	err := json.Unmarshal(payload, &v)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v, nil
+}
+
+// CastVote is a signed vote.
+type CastVote struct {
```

Comment:
The cms v1 api defines the vote as either an "aye" or "nay" but this defines the vote as a hex encoded vote bit.  The v1 api struct naming and cmsplugin naming should be consistent.